File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        if (head == nullptr) { return false; }
        if (head->next == nullptr) { return false; }

        auto slow = head;
        auto fast = head;

        do {
            if (slow == nullptr || fast == nullptr || fast->next == nullptr) {
                return false;
            }

            slow = slow->next;
            fast = fast->next->next;
        } while (slow != fast);

        return true;

    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        length = 0
        temp = head
        
        while temp != None:
            temp = temp.next
            length += 1
        
        pos = 0
        curr, prev = head, None

        while pos < length - n and curr != None:
            prev = curr
            curr = curr.next
            pos += 1
        
        if prev != None:
            if curr != None: # Delete from the middle
                prev.next = curr.next
            else: # Delete from the end
                prev.next = None
        else: # Delete from the front
            head = head.next
        
        return head
